numpy/core/src/umath/ufunc_object.c

```diff
@@ -2089,6 +2107,122 @@ PyUFunc_GeneralizedFunction(PyUFuncObject *ufunc,
+
+        if (axis_list) {
+            if (PyList_Size(axis) != nop) {
+                PyErr_Format(PyExc_ValueError,
+                             "axis should be a list with %d elements", nop);
+                retval = -1;
+                goto fail;
+            }
+        } else {
+            /* a single integer is OK if there is only one core dimension */
+            if (ufunc->core_num_dim_ix > 1) {
+                PyErr_SetString(PyExc_ValueError, "axis should be a list");
+                retval = -1;
+                goto fail;
+            }
+            axis_rqst = (int)PyInt_AsSsize_t(axis);
+            if (error_converting(axis_rqst)) {
+                PyErr_Clear();
+                PyErr_SetString(PyExc_ValueError,
+                                "axis should be a list or a single integer");
+                retval = -1;
+                goto fail;
+            }
+        }
+        for (i = 0; i < nop; ++i) {
+            int op_ndim, iax, rqst, dflt;
+            int num_dims = ufunc->core_num_dims[i];
+            if (op[i] != NULL) {
+                op_ndim = PyArray_NDIM(op[i]);
+            }
+            else {
+                op_ndim = broadcast_ndim + ufunc->core_num_dims[i];
+            }
+            /* initialize remap to straight one to one */
+            for (j = 0; j < op_ndim; j++) {
+                remap_axis[i][j] = j;
+            }
+            if (axis_list) {
+                int n;
+                item = PyList_GET_ITEM(axis, i);
+                if (item == Py_None) { /* no remap for this operand */
+                    continue;
+                }
+                item_list = PyList_Check(item);
+                if (item_list) {
+                    n = PyList_Size(item);
+                }
+                else { /* try as integer */
+                    rqst = (int)PyInt_AsSsize_t(item);
+                    if (error_converting(rqst)) {
+                        PyErr_Clear();
+                        n = -1;  /* to give descriptive error message below */
+                    }
+                    else {
+                        n = 1;
+                    }
+                }
+                if (n != num_dims) {
+                    if (num_dims == 1) {
+                        PyErr_Format(PyExc_ValueError,
+                                     "axis item %d should be a list with a "
+                                     "single element, an integer, or None", i);
+                    }
+                    else {
+                        PyErr_Format(PyExc_ValueError,
+                                     "axis item %d should be a list with %d "
+                                     "elements or None", i, num_dims);
+                    }
+                    retval = -1;
+                    goto fail;
+                }
+            } else {
+                rqst = axis_rqst; /* use same axis for all operands */
+            }
+            for (iax = num_dims - 1; iax >= 0; --iax) {
```

Comment:
I think this logic is unnecessarily complicated. And unless I am missing something, it doesn't check that there are no repeated items in `item`, does it? Rather than shuffling entries in `remap_axis` around, you could fill it directly doing something like:

```
int have_seen_axis[NPY_NUM_DIMS] = {0};

for (j = 0; j < num_dims; ++j) {
    if (item_list) {
        rqst = (int)PyInt_AsSsize_t(PyList_GET_ITEM(item, j));
        if (error_converting(rqst)) {
            retval = -1;
            goto fail;
        }
    }
    if (check_and_adjust_axis(&rqst, op_ndim) < 0) {
        retval = -1;
        goto fail;
    }
    if (have_seen_axis[rqst]) {
        PyErr_Format(PyExc_ValueError,
                     "axis item %d has value %d repeated", i, rqst);
    }
    have_seen_axis[rqst] = 1;
    remap_axis[i][op_ndim - num_dims + j] = rqst;
}
iax = 0;
for (j = 0; j < op_ndim - num_dims; ++j) {
    while (have_seen_axis[iax]) {
        ++iax;
    }
    remap_axis[i][j] = iax;
}
```